lib/hyperion/formats.rb

```diff
@@ -0,0 +1,84 @@
+require 'oj'
+require 'stringio'
+require 'hyperion/aux/logger'
+require 'abstractivator/enum'
+
+class Hyperion
+  module Formats
+    # Serializes and deserializes the supported formats.
+    # This is done as gracefully as possible.
+
+    include Hyperion::Logger
+
+    define_enum :Known, :json, :protobuf
+
+    def write(obj, format)
+      return obj if obj.is_a?(String) || obj.nil? || format.nil?
+
+      case Formats.get_from(format)
+        when :json; write_json(obj)
+        when :protobuf; obj
+        else; fail "Unsupported format: #{format}"
+      end
+    end
+
+    def read(bytes, format)
+      return nil if bytes.nil?
+      return bytes if format.nil?
+
+      case Formats.get_from(format)
+        when :json; read_json(bytes)
+        when :protobuf; bytes
+        else; fail "Unsupported format: #{format}"
+      end
+    end
+
+    def self.get_from(x)
+      x.respond_to?(:format) ? x.format : x
+    end
+
+    private
+
+    def write_json(obj)
+      begin
+        TimeAsJsonShim.hyperion_mode = true
+        Oj.dump(obj, mode: :compat)
+      ensure
+        TimeAsJsonShim.hyperion_mode = false
+      end
+    end
+
+    def read_json(bytes)
+      begin
+        Oj.compat_load(bytes, mode: :compat)
+      rescue Oj::ParseError => e
+        logger.error e.message
+        bytes
+      end
+    end
+
+    def get_oj_line_and_col(e)
+      m = e.message.match(/at line (?<line>\d+), column (?<col>\d+)/)
+      m ? [m[:line].to_i, m[:col].to_i] : nil
+    end
+
+  end
+end
+
+module TimeAsJsonShim
+  mattr_accessor :hyperion_mode
+
+  def as_json(*)
+    if TimeAsJsonShim.hyperion_mode
+      self.utc.iso8601(3)
+    elsif defined? super
+      super
+    else
+      to_s
+    end
+  end
+end
+
+class Time
+  include TimeAsJsonShim
+end
```

Comment:
I understand what you were trying to do with the `TimeAsJsonShim`. You wanted to control how Oj output Time instances as json. It is too bad that Oj does not have a better way to configure this than to shim into the class. 

I found this link that talks of the Oj configuration. I think that if you look at the time_format option and choose ':xmlschema', it may give what you are looking for. http://www.ohler.com/oj/Oj.html

The reason I suggest xmlschema is that the method you are using on time `iso8601()` is an alias for xmlschema on the Time class (standard-lib) http://ruby-doc.org/stdlib-2.2.0/libdoc/time/rdoc/Time.html#method-i-iso8601